packages/js/product-editor/src/components/variations-table/use-variations/use-variations.ts

```diff
@@ -0,0 +1,451 @@
+						if ( options.length === 0 ) {
+							return prev;
+						}
+						return [ ...prev, { ...item, terms: options } ];
+					}
+					return [ ...prev, item ];
+				},
+				[]
+			);
+
+			if ( ! isPresent ) {
+				newFilter.push( {
+					attribute: attribute.slug,
+					terms: options,
+				} );
+			}
+
+			pageRef.current = 1;
+
+			setIsLoading( true );
+			getCurrentVariationsPage( productId, newFilter ).finally( () =>
+				setIsLoading( false )
+			);
+
+			setFilters( newFilter );
+		};
+	}
+
+	function getFilters( attribute: ProductAttribute ) {
+		return (
+			filters.find( ( filter ) => filter.attribute === attribute.slug )
+				?.terms ?? []
+		);
+	}
+
+	function hasFilters() {
+		return filters.length;
+	}
+
+	function clearFilters() {
+		setFilters( [] );
+	}
+
+	// Variation selection
+
+	const [ selectedCount, setSelectedCount ] = useState( 0 );
+	const [ isSelectingAll, setIsSelectingAll ] = useState( false );
+
+	const selectedVariationsRef = useRef< Record< number, ProductVariation > >(
+		{}
+	);
+
+	const selected = useMemo(
+		function getSelected() {
+			return selectedCount > 0
+				? Object.values( selectedVariationsRef.current )
+				: [];
+		},
+		[ selectedCount ]
+	);
+
+	const isSelected = useCallback(
+		function isSelected( variation: ProductVariation ) {
+			return (
+				selectedCount > 0 &&
+				variation.id in selectedVariationsRef.current
+			);
+		},
+		[ selectedCount ]
+	);
+
+	const areAllSelected = useMemo(
+		() => selectedCount > 0 && variations.every( isSelected ),
+		[ variations, selectedCount, isSelected ]
+	);
+
+	const areSomeSelected = useMemo(
+		() => selectedCount > 0 && variations.some( isSelected ),
+		[ variations, selectedCount, isSelected ]
+	);
+
+	function onSelect( variation: ProductVariation ) {
+		return function handleChange( checked: boolean ) {
+			if ( checked ) {
+				selectedVariationsRef.current[ variation.id ] = variation;
+				setSelectedCount( ( current ) => current + 1 );
+			} else {
+				delete selectedVariationsRef.current[ variation.id ];
+				setSelectedCount( ( current ) => current - 1 );
+			}
+		};
+	}
+
+	function onSelectPage( checked: boolean ) {
+		if ( checked ) {
+			variations.forEach( ( variation ) => {
+				selectedVariationsRef.current[ variation.id ] = variation;
+			} );
+		} else {
+			variations.forEach( ( variation ) => {
+				delete selectedVariationsRef.current[ variation.id ];
+			} );
+		}
+		setSelectedCount( Object.keys( selectedVariationsRef.current ).length );
+	}
+
+	async function onSelectAll() {
+		setIsSelectingAll( true );
+
+		const { getProductVariations } = resolveSelect(
+			EXPERIMENTAL_PRODUCT_VARIATIONS_STORE_NAME
+		);
+
+		let currentPage = 1;
+		let fetchedCount = 0;
+
+		while ( fetchedCount < totalCount ) {
+			const chunk = await getProductVariations< ProductVariation[] >( {
+				product_id: productId,
+				page: currentPage++,
+				per_page: 50,
+				order: 'asc',
+				orderby: 'menu_order',
+			} );
```

Comment:
I don't understand this @louwie17

Reply:
Which part?
> One slightly annoying thing about this is that if a user still manually moves pages, we are still requesting the variations for each page, because the per_page is different.

In the above request you default to `per_page` of 50, but the default shown is 5. Meaning that if we go to page 2, it will do a full request to get the next 5. Even if a user might have already selected the `select all` and we retrieved all 50 of them.
This isn't a huge deal though and I don't necessarily want to handle the pagination purely client side to avoid this.

Does that help clarify?

Reply:
Well, the 50 per page is only to select all variations in chunks, of course a new request is going to be made, it has to. The intention here was to reduce the amount of requests while keeping a fare number of items per page.